geomstats/geometry/hypersphere.py

```diff
@@ -357,14 +357,41 @@ def exp(self, tangent_vec, base_point):
         """
         tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=2)
         base_point = gs.to_ndarray(base_point, to_ndim=2)
-        # TODO(johmathe): Evaluate the bias introduced by this variable
-        # TODO(Xavier): this is the wrong way to prevent numerical error:
-        #  if norm_tangent_vec is too small we should use a Taylor expansion.
-        norm_tangent_vec = self.embedding_metric.norm(tangent_vec) + EPSILON
-        coef_1 = gs.cos(norm_tangent_vec)
-        coef_2 = gs.sin(norm_tangent_vec) / norm_tangent_vec
+
+        # TODO(nina): Decide on metric.space or space.metric
+        #  for the hypersphere
+        # TODO(nina): Raise error when vector is not tangent
+        n_base_points, extrinsic_dim = base_point.shape
+        n_tangent_vecs, _ = tangent_vec.shape
+        n_exps = gs.cast(gs.maximum(n_base_points, n_tangent_vecs), gs.int32)
+
+        hypersphere = Hypersphere(dimension=extrinsic_dim-1)
+        proj_tangent_vec = hypersphere.projection_to_tangent_space(
+            tangent_vec, base_point)
+        norm_tangent_vec = self.embedding_metric.norm(proj_tangent_vec)
+
+        n_tiles_vec = gs.cast(gs.divide(n_exps, n_tangent_vecs), gs.int32)
+        norm_tangent_vec = gs.tile(norm_tangent_vec, [n_tiles_vec, 1])
+        n_tiles_base_point = gs.cast(
+            gs.divide(n_exps, n_base_points), gs.int32)
+        base_point = gs.tile(base_point, [n_tiles_base_point, 1])
+
+        mask_0 = gs.isclose(norm_tangent_vec, 0.)
+        mask_non0 = ~mask_0
+
+        coef_1 = gs.zeros((n_exps, 1))
+        coef_2 = gs.zeros((n_exps, 1))
+        norm2 = norm_tangent_vec[mask_0]**2
+        norm4 = norm2**2
+        norm6 = norm2**3
+        coef_1[mask_0] = 1. - norm2/2. + norm4/24. - norm6/720.
```

Comment:
Sorry to revive this after the PR was already merged, but for clarity these constants should probably be kept in a look-up table like https://github.com/geomstats/geomstats/blob/c95984472e9e91eec3eed455a95d1d62132d02a1/geomstats/geometry/hypersphere.py#L18 In fact, `COS_TAYLOR_COEFFS` is nowhere used in the module at all atm. I'm surprised flake8 doesn't catch this...

Reply:
Creating an issue now